sycl/test/check_device_code/address_space_cast.cpp

```diff
@@ -0,0 +1,43 @@
+// This test checks that the dynamic address casts can be optimized out
+// (via LowerSYCLAddressSpaceCasts) when the address spaces are known at compile
+// time by checking that only one of from {foo, bar, baz} are called. Note:
+// this can only happen if dispatch is inlined, which is why we use
+// __attribute__((always_inline)) on it.
+
+// XFAIL: *
+// RUN: %clangxx -O1 -DADDR_SPACE=global_space -S -emit-llvm -fsycl -fsycl-device-only %s -o - \
+// RUN: | FileCheck %s --implicit-check-not=bar --implicit-check-not=baz
+// RUN: %clangxx -O1 -DADDR_SPACE=local_space -S -emit-llvm -fsycl -fsycl-device-only %s -o - \
+// RUN: | FileCheck %s --implicit-check-not=foo --implicit-check-not=baz
+// RUN: %clangxx -O1 -DADDR_SPACE=private_space -S -emit-llvm -fsycl -fsycl-device-only %s -o - \
+// RUN: | FileCheck %s --implicit-check-not=foo --implicit-check-not=bar
+
+#include <sycl/ext/oneapi/experimental/address_cast.hpp>
+#include <sycl/multi_ptr.hpp>
+
+SYCL_EXTERNAL float foo(float x);
+SYCL_EXTERNAL float bar(float x);
+SYCL_EXTERNAL float baz(float x);
```

Comment:
Let's use more descriptive function names.
```suggestion
SYCL_EXTERNAL float globalAS(float x);
SYCL_EXTERNAL float localAS(float x);
SYCL_EXTERNAL float privateAS(float x);
```